### PR TITLE
Mellanox driver updates

### PR DESCRIPTION
**Pull request by ktbyers**

### CODE REVIEW

File: netmiko/mellanox/mellanox_mlnxos_ssh.py [hunk 1]
```diff
@@ -42,7 +67,7 @@ def exit_config_mode(self, exit_config="exit", pattern="#"):
         if self.check_config_mode():
             self.write_channel(self.normalize_cmd(exit_config))
             output = self.read_until_pattern(pattern=pattern)
-            if self.check_config_mode():
+            if not self.check_enable_mode():
```
Comment (ktbyers) [inline netmiko/mellanox/mellanox_mlnxos_ssh.py]: That looks incorrect i.e. at that point we should be checking that we are no longer in config_mode (not whether we are in enable_mode). Was there some reason this was required?

I am probably going to strip this out and just call the `exit_config_mode` in the parent classes (i.e. use `super` on it).

File: netmiko/mellanox/mellanox_mlnxos_ssh.py [hunk 2]
```diff
@@ -1,10 +1,35 @@
...
+        return super(MellanoxMlnxosSSH, self).check_enable_mode(
+            check_string=check_string
+        )
+
+    def enable(self, ena_cmd="enable", pattern="#"):
+        """Enter into enable mode."""
+        output = ""
+        if not self.check_enable_mode():
+            self.write_channel(self.normalize_cmd(ena_cmd))
+            output = self.read_until_pattern(pattern=pattern)
```
Comment (ktbyers) [inline netmiko/mellanox/mellanox_mlnxos_ssh.py]: I see you never send the `secret` so does Mellanox just transition to `enable` mode with no additional password?

**Comment by ktbyers** [comment]
@vasorokin See my comments I made above. I am going to submit some updates to this driver in ten minutes or so.

**Comment by ktbyers** [comment]
@vasorokin @juliakreger FYI, I made some updates to the driver. 

I consolidated the driver into just the new driver i.e. both `mellanox` and `mellanox_mlnxos` in SSH dispatcher will point to this new driver and old Mellanox driver was removed. I checked the history and the old driver was intended for Mellanox Switch. There is a possibility here that I broke the old code in which case I will just have to make fixes for older Mellanox devices. 

I consolidated a bunch of the calls in the new driver to just use `super` calls to the parent classes or to directly rely on inheritance in cases where the behavior was identical to the default cisco behavior.

It would be good if you could re-test this updated driver works in your test scenario / works with your requirements.

Just let me know if I broke something with my changes.

File: netmiko/mellanox/mellanox_mlnxos_ssh.py (hunk 2 above)
Comment (vasorokin) [reply netmiko/mellanox/mellanox_mlnxos_ssh.py]: It does not ask me for password

File: netmiko/mellanox/mellanox_mlnxos_ssh.py (hunk 1 above)
Comment (vasorokin) [reply netmiko/mellanox/mellanox_mlnxos_ssh.py]: Yes that probably makes sense, please go ahead change it.

**Comment by ktbyers** [comment]
@vasorokin I am going to go ahead and merge this. If there are issues with the driver, please let me know.

Thanks, Kirk

**Comment by ktbyers** [comment]
@vasorokin Any possibility you can re-run the Netmiko unit tests using this new Netmiko driver (against a Mellanox device)?

I am a bit reluctant to merge this in without regression testing it against a device.

Netmiko testing document is here:

https://github.com/ktbyers/netmiko/blob/develop/TESTING.md

It would be good to run the `config` and `show` tests against a device.

**Comment by DragoonAethis** [comment]
I've used a forked version of this driver for a bit and it worked fine, but I can retest with your branch tomorrow/Wednesday, if that sounds fine for you? (Admittedly, I'd very much like to see this in upstream packages :)

**Comment by ktbyers** [comment]
@DragoonAethis Yes, if you could re-test it that would be great. I want to make sure I didn't break anything with the changes/updates to this driver.

Thanks, Kirk

**Comment by vasorokin** [comment]
@ktbyers
Was testing with the real switchport assignment to VLAN from Ironic instead of https://github.com/ktbyers/netmiko/blob/develop/TESTING.md
Got the following error in Ironic's server.log:

2019-07-02 10:39:41.381 3742 ERROR neutron.plugins.ml2.managers [req-61fce289-b6ef-4aed-a5d6-d9ce8143141a 74ff7c27f72c4b8881ecbad6420ad35d b57ded455d3b4e458aab759dcfa6836b - default default] Mechanism driver genericswitch failed in bind_port: GenericSwitchNetmikoConnectError: Netmiko connection error: {'username': 'swadmin', 'ip': '192.168.32.42', 'password': '******', 'device_type': 'mellanox_mlnxos'}, error: Failed to exit configuration mode

2019-07-02 10:39:41.381 3742 ERROR neutron.plugins.ml2.managers Traceback (most recent call last):
2019-07-02 10:39:41.381 3742 ERROR neutron.plugins.ml2.managers   File "/usr/lib/python2.7/site-packages/neutron/plugins/ml2/managers.py", line 783, in _bind_port_level
2019-07-02 10:39:41.381 3742 ERROR neutron.plugins.ml2.managers     driver.obj.bind_port(context)
2019-07-02 10:39:41.381 3742 ERROR neutron.plugins.ml2.managers   File "/usr/lib/python2.7/site-packages/networking_generic_switch/generic_switch_mech.py", line 472, in bind_port
2019-07-02 10:39:41.381 3742 ERROR neutron.plugins.ml2.managers     switch.plug_port_to_network(port_id, segmentation_id)
2019-07-02 10:39:41.381 3742 ERROR neutron.plugins.ml2.managers   File "/usr/lib/python2.7/site-packages/networking_generic_switch/devices/netmiko_devices/__init__.py", line 193, in plug_port_to_network
2019-07-02 10:39:41.381 3742 ERROR neutron.plugins.ml2.managers     self.send_commands_to_device(cmds)
2019-07-02 10:39:41.381 3742 ERROR neutron.plugins.ml2.managers   File "/usr/lib/python2.7/site-packages/networking_generic_switch/devices/netmiko_devices/__init__.py", line 149, in send_commands_to_device
2019-07-02 10:39:41.381 3742 ERROR neutron.plugins.ml2.managers     config=device_utils.sanitise_config(self.config), error=e)
2019-07-02 10:39:41.381 3742 ERROR neutron.plugins.ml2.managers GenericSwitchNetmikoConnectError: Netmiko connection error: {'username': 'swadmin', 'ip': '192.168.32.42', 'password': '******', 'device_type': 'mellanox_mlnxos'}, error: Failed to exit configuration mode
2019-07-02 10:39:41.381 3742 ERROR neutron.plugins.ml2.managers 
2019-07-02 10:39:41.430 3742 ERROR neutron.plugins.ml2.managers [req-61fce289-b6ef-4aed-a5d6-d9ce8143141a 74ff7c27f72c4b8881ecbad6420ad35d b57ded455d3b4e458aab759dcfa6836b - default default] Failed to bind port f5197f18-efd4-4f72-96e4-009cae8af43a on host a1497eaf-ac48-4f2c-bedd-0800e0335636 for vnic_type baremetal using segments [{'network_id': 'b42e185c-c963-49a0-872f-abfb40a979fc', 'segmentation_id': 401, 'physical_network': u'10G-net', 'id': 'a6fa4bd4-04af-4095-97ca-02c1405ebaa1', 'network_type': u'vlan'}]
2019-07-02 10:39:41.431 3742 INFO neutron.plugins.ml2.plugin [req-61fce289-b6ef-4aed-a5d6-d9ce8143141a 74ff7c27f72c4b8881ecbad6420ad35d b57ded455d3b4e458aab759dcfa6836b - default default] Attempt 2 to bind port f5197f18-efd4-4f72-96e4-009cae8af43a
2019-07-02 10:39:41.523 3742 INFO paramiko.transport [-] Connected (version 2.0, client OpenSSH_7.1)
2019-07-02 10:39:41.697 3742 INFO paramiko.transport [-] Auth banner: 
Mellanox MLNX-OS Switch Management

So I think I did check_enable_mode() for a reason. Here's the sample command flow with prompts:
10GleafR2 [standalone: master] > 
10GleafR2 [standalone: master] > ena
10GleafR2 [standalone: master] # conf t
10GleafR2 [standalone: master] (config) # sh bios
BIOS version : 4.6.5
10GleafR2 [standalone: master] (config) # exit
10GleafR2 [standalone: master] # exit
Connection to 192.168.32.42 closed.
$

**Comment by ktbyers** [comment]
@vasorokin Does Mellanox support `end` for completely exiting config mode. For example, on Cisco IOS and IOS like devices if you go multiple levels into the config hierarchy, then `exit` will only take you back one level whereas `end` will take you completely out of config mode.

I wonder if we need to change this line to `end`?

```python
 def exit_config_mode(self, exit_config="exit", pattern="#"):
```

Your tests above are really hard to decipher from a netmiko perspective as your error handling hides all the information from where the error occurred in Netmiko and also provides very little information on what you were doing when the error happened.

What configuration changes were you making when the error occurred? Can you reproduce this error in a simple python script outside of your test suite?


Regards,

Kirk

**Comment by vasorokin** [comment]
@ktbyers 
No it does not support 'end':
10GleafR2 [standalone: master] (config interface ethernet 1/1) # end
% Unrecognized command "end".
Type "?" for help.
10GleafR2 [standalone: master] (config interface ethernet 1/1) # 

The configuration change was in assigning a switchport to a VLAN

I'll create the test from the python script or from the same Ironic with the extended (DEBUG?) setting which will be showing the errors and commands in Netmiko. I can do that next week, is that OK for you?

**Comment by DragoonAethis** [comment]
I've retested the driver and it *mostly* works, with the following notes that I've missed previously or patched and forgot about that:
- Terminal paging can be disabled for arbitrary number of lines with `no cli session paging enable` (`show interfaces` plopped over 1000 lines of output, which dropped into the pager)
- Mellanox prompt doesn't support `end`, but exiting config mode can be done by issuing `exit` until the prompt ends with `(config) #`, then one more time to actually drop out of the config mode. (That one I've missed due to always issuing full commands without landing anywhere deeper in the config hierarchy...)

Other than that, this works fine:
```python
from netmiko import Netmiko
nm = Netmiko("amazing.hostname", username="admin", password=":DDDD", device_type="mellanox_mlnxos")
nm.send_command("no cli session paging enable")
nm.enable()
nm.send_config_set(['interface ethernet 1/30 shutdown', 'interface ethernet 1/30 speed auto', 'no interface ethernet 1/30 shutdown'])
nm.disconnect()```

**Comment by ktbyers** [comment]
Okay, bummer Mellanox doesn't have a good way of just exiting config mode on a single command...so yes, we will have to do as you describe, keep typing `exit` until `(config` is gone.

Sounds good on the better paging disable command.

**Comment by ktbyers** [comment]
I suspect the lack of `end` issue is the same issue that @vasorokin ran into...so fixing that will hopefully fix his issue.

**Comment by vasorokin** [comment]
[netmiko-mellanox.log](https://github.com/ktbyers/netmiko/files/3369773/netmiko-mellanox.log)
@ktbyers please see attached netmiko entries from the Ironic debug log
The configuration change was in assigning a switchport to a VLAN

**Comment by ktbyers** [comment]
FYI, I updated this PR to fix both the exit config mode issue and the paging issue. It would be good if you could re-test this new code. Thanks.

@vasorokin

**Comment by vasorokin** [comment]
That worked! Thank you @ktbyers 
Here is the log if you interested in.
[netmiko-mellanox.log](https://github.com/ktbyers/netmiko/files/3386636/netmiko-mellanox.log)

**Comment by ktbyers** [comment]
Okay, I will also let @DragoonAethis test this in the next few days. Probably try to merge it next week if we can.


Regards, Kirk

**Comment by DragoonAethis** [comment]
I'll test it next Monday/Tuesday.

**Comment by DragoonAethis** [comment]
That's not quite the Monday I've meant a week ago, but tested - everything works correctly, can be merged.

**Pull request merged**